nixos/modules/services/networking/crab-hole.nix

```
@@ -0,0 +1,230 @@
+              }
+              {
+                protocol = "tls";
+                socket_addr = "1.1.1.1:853";
+                tls_dns_name = "1dot1dot1dot1.cloudflare-dns.com";
+                trust_nx_responses = false;
+              }
+            ];
+            options = {
+              validate = false;
+            };
+          };
+        };
+
+        type = lib.types.submodule {
+          freeformType = settingsFormat.type;
+          options = {
+            blocklist = {
+              include_subdomains = lib.mkEnableOption "Include Subdomains";
+              lists = lib.mkOption {
+                type = lib.types.listOf (lib.types.either lib.types.str lib.types.path);
+                default = [ ];
+                description = "List of blocklists. If files are added via url, make sure the service has access to them!";
+                apply = x: (lib.lists.forEach x (v: if builtins.isPath v then ("file://${v}") else v));
+              };
+              allow_list = lib.mkOption {
+                type = lib.types.listOf (lib.types.either lib.types.str lib.types.path);
+                default = [ ];
+                description = "List of allowlists. If files are added via url, make sure the service has access to them!";
+                apply = x: (lib.lists.forEach x (v: if builtins.isPath v then ("file://${v}") else v));
+              };
+            };
+
+            downstream = lib.mkOption {
+              description = "Configuration for the downstream DNS Servers";
+              type = lib.types.listOf (
+                lib.types.submodule {
+                  freeformType = settingsFormat.type;
+                  options = {
+                    protocol = lib.mkOption {
+                      type = lib.types.enum [
+                        "udp"
+                        "tls"
+                        "https"
+                        "quic"
+                      ];
+                      example = "udp";
+                      default = "udp";
+                      description = "The protocol to listen on";
+                    };
+                    listen = lib.mkOption {
+                      type = lib.types.str;
+                      example = "[::]";
+                      default = "localhost";
+                      description = "The IP to listen on";
+                    };
+                    port = lib.mkOption {
+                      type = lib.types.ints.u16;
+                      example = 53;
+                      default = 53;
+                      description = "The port to listen on";
+                    };
+                  };
+                }
+              );
+            };
+
+            upstream = {
+              options = lib.mkOption {
+                default = { };
+                description = "Set resolover options. For a full list see https://docs.rs/trust-dns-resolver/0.23.0/trust_dns_resolver/config/struct.ResolverOpts.html";
```

Typo. Also linking a fixed version of the Rust doc feels iffy
```suggestion
                description = ''
                	Resolver options.
                	For a full list see <https://docs.rs/trust-dns-resolver/0.23.0/trust_dns_resolver/config/struct.ResolverOpts.html>
                '';
```

---

Given that crab-hole depends on a specific version of hickory-dns, currently 0.24, I would suggest to replace this link with <https://docs.rs/hickory-resolver/0.24/hickory_resolver/config/struct.ResolverOpts.html>. I don't think using `latest` would be a good idea since the options can change in newer version of hickory-dns before crab-hole updates to that version of hickory-dns. That of course means that the link might need updating with new releases of crab-hole, and at least we've failed to do that in our own readme where this link probably comes from :see_no_evil: